### PR TITLE
cmd/fetch-payload: a go command to fetch payload

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

This is meant to replace `./hack/fetch-releases.sh` in a more flexible
way. In a gist, you can use different components yaml and fetch from
them depending on the need.

On additional goal is to be able to support fetch from PRs and/or
local folders.

```bash
$ go run ./cmd/fetch-payload -filename components.yaml -target foo/ -platform openshift
skip: component tekton-dasbhoard on platform openshift
Download https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml to foo/cmd/openshift/kodata/tekton-pipeline/0.0.0-latest/00-release.yaml
Download https://raw.githubusercontent.com/openshift-pipelines/pipelines-as-code/release-0.5.3/release-0.5.3.yaml to foo/cmd/openshift/kodata/tekton-addon/pipelines-as-code/0.5.3/00-release.yaml
skip: component tekton-results on platform openshift
Download https://storage.googleapis.com/tekton-releases/chains/latest/release.yaml to foo/cmd/openshift/kodata/tekton-chains/0.0.0-latest/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/triggers/latest/interceptors.yaml to foo/cmd/openshift/kodata/tekton-trigger/0.0.0-latest/01-interceptors.yaml
Download https://storage.googleapis.com/tekton-releases/triggers/latest/release.yaml to foo/cmd/openshift/kodata/tekton-trigger/0.0.0-latest/00-release.yaml

$ go run ./cmd/fetch-payload -filename components.yaml -target foo/
Download https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml to foo/cmd/kubernetes/kodata/tekton-pipeline/0.0.0-latest/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/triggers/latest/interceptors.yaml to foo/cmd/kubernetes/kodata/tekton-trigger/0.0.0-latest/01-interceptors.yaml
Download https://storage.googleapis.com/tekton-releases/chains/latest/release.yaml to foo/cmd/kubernetes/kodata/tekton-chains/0.0.0-latest/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/triggers/latest/release.yaml to foo/cmd/kubernetes/kodata/tekton-trigger/0.0.0-latest/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/results/previous/v0.4.0/release.yaml to foo/cmd/kubernetes/kodata/tekton-results/v0.4.0/00-release.yaml
Download https://raw.githubusercontent.com/openshift-pipelines/pipelines-as-code/release-0.5.3/release-0.5.3.yaml to foo/cmd/kubernetes/kodata/tekton-addon/pipelines-as-code/0.5.3/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/dashboard/latest/tekton-dashboard-release.yaml to foo/cmd/kubernetes/kodata/tekton-dasbhoard/tekton-dashboard-fullaccess/0.0.0-latest/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/dashboard/latest/tekton-dashboard-release-readonly.yaml to foo/cmd/kubernetes/kodata/tekton-dasbhoard/tekton-dashboard-readonly/0.0.0-latest/01-release-readonly.yaml

$ go run ./cmd/fetch-payload -filename components.local.yaml -target foo/
Copy ../results/resolved.yaml to foo/cmd/kubernetes/kodata/tekton-results/v0.4.0/00-release.yaml
Copy ../chains/resolved.yaml to foo/cmd/kubernetes/kodata/tekton-chains/0.0.0-latest/00-release.yaml
Copy ../pipeline/resolved.yaml to foo/cmd/kubernetes/kodata/tekton-pipeline/0.0.0-latest/00-release.yaml
Copy ../triggers/resolved.release.yaml to foo/cmd/kubernetes/kodata/tekton-trigger/0.0.0-latest/00-release.yaml
Copy ../triggers/resolved.interceptors.yaml to foo/cmd/kubernetes/kodata/tekton-trigger/0.0.0-latest/01-interceptors.yaml
Download https://storage.googleapis.com/tekton-releases/dashboard/latest/tekton-dashboard-release-readonly.yaml to foo/cmd/kubernetes/kodata/tekton-dasbhoard/tekton-dashboard-readonly/0.0.0-latest/01-release-readonly.yaml
Download https://raw.githubusercontent.com/openshift-pipelines/pipelines-as-code/release-0.5.3/release-0.5.3.yaml to foo/cmd/kubernetes/kodata/tekton-addon/pipelines-as-code/0.5.3/00-release.yaml
Download https://storage.googleapis.com/tekton-releases/dashboard/latest/tekton-dashboard-release.yaml to foo/cmd/kubernetes/kodata/tekton-dasbhoard/tekton-dashboard-fullaccess/0.0.0-latest/00-release.yaml
```

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by vdemeester** [comment]
/hold
Let's wait for 0.55 to get out before merging this.

**Comment by nikhil-thomas** [comment]
> On additional goal is to be able to support fetch from PRs and/or
local folders.

🕺
/meow

File: cmd/fetch-payload/main.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+var (
+	filename  string
+	targetDir string
+	platform  string
+)
+
+func main() {
+	flag.StringVar(&filename, "filename", "", "components configuration to load")
```
Comment (nikhil-thomas) [inline cmd/fetch-payload/main.go:24]: suggestion: `s/filename/config-file/` 🤓

File: cmd/fetch-payload/main.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	return g.Wait()
+}
+
+func fetchComponent(ctx context.Context, name string, c component, targetDir, platform string) error {
+	// fmt.Printf("component: %s\n", name)
+	// fmt.Printf("version: %s\n", c.Version)
+	if !c.toFetch(platform) {
+		fmt.Fprintf(os.Stderr, "skip: component %s on platform %s\n", name, platform)
+		return nil
+	}
```
Comment (nikhil-thomas) [inline cmd/fetch-payload/main.go:67]: 👍

**Comment by nikhil-thomas** [comment]
very valuable and interesting patch. 🧑‍💻👍

comments:
- shall we rename the `components.*.yaml` files to something like `tekton-component-releases.*.yaml` or `tekton-release-dependencies.*.yaml` or `tekton-requirements.yaml` 🤓 
- could you add a few unit tests so that anyone could contribute/maintain this without worrying about regression bugs. 🧑‍💻

**Comment by vdemeester** [comment]
@nikhil-thomas should we also add addon/cluster task support in ?

**Comment by nikhil-thomas** [comment]
> @nikhil-thomas should we also add addon/cluster task support in ?

that will be wonderful 💯 👍

**Comment by vdemeester** [comment]
I need to rework this a bit…

**Comment by nikhil-thomas** [comment]
/retest

as https://github.com/tektoncd/operator/pull/769 has merged, ci should pass

**Comment by vdemeester** [comment]
Closing in favor of a smaller, more incremental approach, see https://github.com/tektoncd/operator/pull/898

**Pull request closed without merging**